### ISSUE TITLE
Limit on the number of reference sequences ?

### ISSUE THREAD

**Issue by alxsimon** [nkschaefer/ad-libs#3]
Hi,

I encountered what I think is a limit on the number of reference sequences that can be provided.

When including more than 10 reference sequences for one of the population (A or B), I have the following error message:
```
Traceback (most recent call last):
  File "/opt/ad-libs/bin/adlibs_hmm.py", line 450, in <module> 
    sys.exit(main(sys.argv))
  File "/opt/ad-libs/bin/adlibs_hmm.py", line 420, in main 
    query_data = win_scores[:,3].astype('float')
IndexError: too many indices for array 
```

At first I thought this had to do with some fasta files being badly formatted but when shuffling the reference sequencing between all the sequences I have but staying below the 10 limit, this is working perfectly well.

Here is the config file with which the error appears
```
# gallo_med reference sequences
A output/fasta_seq/Mgallo/1642-AF-107.sub output/fasta_seq/Mgallo/1642-AF-108.sub output/fasta_seq/Mgallo/1642-AF-109.sub output/fasta_seq/Mgallo/1642-AF-110.sub output/fasta_seq/Mgallo/1642-AF-111.sub output/fasta_seq/Mgallo/1642-AF-112.sub output/fasta_seq/Novaseq/Her-19.sub output/fasta_seq/Hiseq/Her-17.sub output/fasta_seq/Novaseq/Th16-04.sub output/fasta_seq/Hiseq/Th16-07.sub output/fasta_seq/Hiseq/Bizerte-19.sub output/fasta_seq/Hiseq/Bizerte-23.sub
# edu_eu reference sequences
B output/fasta_seq/Hiseq/Holl-01.sub output/fasta_seq/Hiseq/Holl-02.sub output/fasta_seq/Novaseq/Ouis-01.sub output/fasta_seq/Hiseq/Ouis-02.sub output/fasta_seq/Novaseq/Revi-01.sub output/fasta_seq/Hiseq/Revi-02.sub
# Window size in bp
WINDOW 5000
# Estimated admixed population size
N 3000
# Estimated generations since admixture
g 40
# What percent of a window must be "N" nucleotides before it's ignored?
SKIP 0.75
# How many processes should be run at a time (max. 1 per hybrid genome)?
PROCS 12
# Prefix of output files (including directory)
OUTPUT output/adlibs/adlibs_havre
# Nucleotide diversity values (in population A, in population B, between pops A & B)
PI 0.001 0.001 0.03
# List all hybrid individuals, then name, and optionally sex and percent population A
# ancestry (genome-wide). If this number is not known, it can be calculated, as long
# as an outgroup sequence (specified with OUTGROUP) is provided
# Sex only matters if X chromosome sequences are specified (specified with X):
# X scaffoldX1 scaffoldX2 scaffoldX3
HYBRID output/fasta_seq/Novaseq/Havre-B-08.sub Havre-B-08 M 0.6
HYBRID output/fasta_seq/Novaseq/Havre-G-13.sub Havre-G-13 M 0.6
HYBRID output/fasta_seq/Novaseq/Havre-C-06.sub Havre-C-06 M 0.6
HYBRID output/fasta_seq/Novaseq/Havre-F-11.sub Havre-F-11 M 0.6
HYBRID output/fasta_seq/Novaseq/Havre-J-08.sub Havre-J-08 M 0.6
HYBRID output/fasta_seq/Novaseq/Havre-E-05.sub Havre-E-05 M 0.6
HYBRID output/fasta_seq/Novaseq/LeHaP4-08.sub LeHaP4-08 M 0.6
HYBRID output/fasta_seq/Novaseq/LeHaP11-13.sub LeHaP11-13 M 0.6
HYBRID output/fasta_seq/Novaseq/LeHaP2-03.sub LeHaP2-03 M 0.6
HYBRID output/fasta_seq/Novaseq/LeHaP9-13.sub LeHaP9-13 M 0.6
HYBRID output/fasta_seq/Novaseq/LeHaP3-10.sub LeHaP3-10 M 0.6
HYBRID output/fasta_seq/Novaseq/B-amont-02.sub B-amont-02 M 0.6
```

**Comment by alxsimon** [comment]
I do not know if this is linked but sometimes, not consistently, for some length of sequences it is also throwing pointer errors:
`free(): invalid pointer` or sometimes `double free or corruption (!prev)`

**Comment by alxsimon** [comment]
More information on this issue. I pinpointed the problem to the `adlibs_score` program running it directly:
```
ad-libs/bin/adlibs_score \
-1 output/fasta_seq/Mgallo/1642-AF-107.sub \
-1 output/fasta_seq/Mgallo/1642-AF-108.sub \
-1 output/fasta_seq/Mgallo/1642-AF-109.sub \
-1 output/fasta_seq/Mgallo/1642-AF-110.sub \
-1 output/fasta_seq/Mgallo/1642-AF-111.sub \
-1 output/fasta_seq/Mgallo/1642-AF-112.sub \
-1 output/fasta_seq/Novaseq/Her-19.sub \
-1 output/fasta_seq/Hiseq/Her-17.sub \
-1 output/fasta_seq/Novaseq/Th16-04.sub \
-1 output/fasta_seq/Hiseq/Th16-07.sub \
-1 output/fasta_seq/Hiseq/Bizerte-19.sub \
-1 output/fasta_seq/Hiseq/Bizerte-23.sub \
-2 output/fasta_seq/Hiseq/Holl-01.sub \
-2 output/fasta_seq/Hiseq/Holl-02.sub \
-2 output/fasta_seq/Novaseq/Ouis-01.sub \
-2 output/fasta_seq/Hiseq/Ouis-02.sub \
-2 output/fasta_seq/Novaseq/Revi-01.sub \
-2 output/fasta_seq/Hiseq/Revi-02.sub \
-w 5000 \
-h output/fasta_seq/Novaseq/Havre-B-08.sub \
-s 0.75
```

which gives the dump
```
free(): invalid pointer
Aborted                 (core dumped)
```

If I remove 2 individuals from the -1 population, it works perfectly fine.
Hope that helps.

**Comment by alxsimon** [comment]
Last addition, a valgrind output showing the possible memory leaks
[valgrind-out.txt](https://github.com/nkschaefer/ad-libs/files/3551053/valgrind-out.txt)

which give this output when not throwing an error with only 10 ref individuals
[valgrind-out-normal.txt](https://github.com/nkschaefer/ad-libs/files/3551067/valgrind-out-normal.txt)

Thanks for your help
--Alexis

**Comment by alxsimon** [comment]
I just realised there is an alternative `adlibs_score.py` script provided that works fine with more than 10 individuals. I will use this one in the meantime.

**Comment by nkschaefer** [comment]
Hey - wow, thanks for checking in on the memory leaks and sorry for an incredibly long delay on my part.

This was one of my first C projects, so I had made some boneheaded mistakes that you caught, memory leaks included. These are all fixed now, as is the limit of 10 reference genomes per population (it's now bumped up to 100). I've also added progress messages and updated everything to use Python3. Hope this helps somebody else in the future.
